src/material/datepicker/calendar.ts

```diff
@@ -149,8 +162,21 @@ export class MatCalendarHeader<D> {
       return this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2);
     }
     // Otherwise we are in 'multi-year' view.
-    return Math.floor(this._dateAdapter.getYear(date1) / yearsPerPage) ==
-        Math.floor(this._dateAdapter.getYear(date2) / yearsPerPage);
+    let minYear = 0;
+    if (this.calendar.minDate) {
+      minYear = this._dateAdapter.getYear(this.calendar.minDate);
+    }
+    if (this.calendar.maxDate) {
+      let maxYear = this._dateAdapter.getYear(this.calendar.maxDate);
+      minYear = maxYear - yearsPerPage + 1;
+    }
+    return Math.floor( (this._dateAdapter.getYear(date1) -  minYear) / yearsPerPage) ==
+        Math.floor( (this._dateAdapter.getYear(date2) - minYear) / yearsPerPage);
```

Comment:
[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/floor) seems to indicate that `Math.floor` will always return a `number`. We should probably use `===` here.